plutus-core/executables/Common.hs

```diff
@@ -389,115 +536,143 @@ plutusOpts = hsubparser (
+getInput StdInput         = getContents
+
+-- | Read and parse a source program
+parseInput ::
+  (Executable a, PLC.Rename (a PLC.AlexPosn) ) =>
+  -- | The source program
+  Input ->
+  -- | The output is either a UPLC or PLC program with annotation
+  IO (a PLC.AlexPosn)
+parseInput inp = do
+    bsContents <- BSL.fromStrict . encodeUtf8 . T.pack <$> getInput inp
+    -- parse the UPLC program
+    case PLC.runQuoteT $ parseProgram bsContents of
+      -- when fail, pretty print the parse errors.
+      Left (err :: PLC.ParseError PLC.AlexPosn) ->
+        errorWithoutStackTrace $ PP.render $ pretty err
+      -- otherwise,
+      Right p -> do
+        -- run @rename@ through the program
+        renamed <- PLC.runQuoteT $ rename p
+        -- check the program for @UniqueError@'s
+        let checked = through (Common.checkProgram (const True)) renamed
+        case checked of
+          -- pretty print the error
+          Left (err :: PLC.UniqueError PLC.AlexPosn) ->
+            errorWithoutStackTrace $ PP.render $ pretty err
+          -- if there's no errors, return the parsed program
+          Right _ -> pure p
 
--- | Convert an untyped program to one where the 'name' type is de Bruijn indices.
-toDeBruijn :: UntypedProgram a -> IO (UntypedProgramDeBruijn a)
-toDeBruijn prog =
-  case runExcept @UPLC.FreeVariableError (UPLC.deBruijnProgram prog) of
-    Left e  -> errorWithoutStackTrace $ show e
-    Right p -> return $ UPLC.programMapNames (\(UPLC.NamedDeBruijn _ ix) -> UPLC.DeBruijn ix) p
+-- Read a binary-encoded file (eg, CBOR- or Flat-encoded PLC)
+getBinaryInput :: Input -> IO BSL.ByteString
+getBinaryInput StdInput         = BSL.getContents
+getBinaryInput (FileInput file) = BSL.readFile file
 
+---------------- Name conversions ----------------
+
+-- | Untyped AST with names consisting solely of De Bruijn indices. This is
+-- currently only used for intermediate values during CBOR/Flat
+-- serialisation/deserialisation.  We may wish to add TypedProgramDeBruijn as
+-- well if we modify the CEK machine to run directly on de Bruijnified ASTs, but
+-- support for this is lacking elsewhere at the moment.
+-- | Untyped AST with names consisting solely of De Bruijn indices. This is
+-- currently only used for intermediate values during CBOR/Flat
+-- serialisation/deserialisation.  We may wish to add TypedProgramDeBruijn as
+-- well if we modify the CEK machine to run directly on de Bruijnified ASTs, but
+-- support for this is lacking elsewhere at the moment.
+type UntypedProgramDeBruijn a = UPLC.Program UPLC.DeBruijn PLC.DefaultUni PLC.DefaultFun a
 
 -- | Convert an untyped de-Bruijn-indexed program to one with standard names.
 -- We have nothing to base the names on, so every variable is named "v" (but
 -- with a Unique for disambiguation).  Again, we don't support typed programs.
-fromDeBruijn :: UntypedProgramDeBruijn a -> IO (UntypedProgram a)
+fromDeBruijn :: UntypedProgramDeBruijn a -> IO (UplcProg a)
```

Comment:
It's because everything is in IO atm. But good point. Should I separate many the functions out from the IO ops? Eg `loadPlcASTfromCBOR` etc. @kwxm I shouldn't run in to problems if I did that?

Reply:
I'm not sure why that's in IO.  Maybe it had to be at some point in the past but now doesn't need to be?  You could try making it pure and see what happens.

Reply:
I think at this point it's not worth taking everything out of IO - I want to merge and move on. If it becomes more of an annoyance later we can do that refactoring.